flow/connectors/mongo/mongo.go

```diff
@@ -0,0 +1,431 @@
+			cancelTimeout()
+		}
+	}()
+	addRecord := func(ctx context.Context, record model.Record[model.RecordItems]) error {
+		recordCount += 1
+		if err := req.RecordStream.AddRecord(ctx, record); err != nil {
+			return err
+		}
+		if recordCount == 1 {
+			req.RecordStream.SignalAsNotEmpty()
+			// after the first record, we switch to a timeout context
+			getCtx, cancelTimeout = context.WithTimeout(ctx, req.IdleTimeout)
+		}
+		return nil
+	}
+
+	for recordCount < req.MaxBatchSize && changeStream.Next(getCtx) {
+		var changeDoc bson.M
+		if err := changeStream.Decode(&changeDoc); err != nil {
+			return fmt.Errorf("failed to decode change stream document: %w", err)
+		}
+
+		if operationType, ok := changeDoc["operationType"]; !ok {
+			c.logger.Warn("operationType field not found")
+			continue
+		} else if operationType != "insert" && operationType != "update" && operationType != "replace" && operationType != "delete" {
+			continue
+		}
+
+		clusterTime := changeDoc["clusterTime"].(bson.Timestamp)
+		clusterTimeNanos := time.Unix(int64(clusterTime.T), 0).UnixNano()
+
+		sourceTableName := changeDoc["ns"].(bson.D)[0].Value.(string) + "." + changeDoc["ns"].(bson.D)[1].Value.(string)
+		destinationTableName := req.TableNameMapping[sourceTableName].Name
+
+		items := model.NewRecordItems(2)
+
+		if documentKey, found := changeDoc["documentKey"]; found {
+			if len(documentKey.(bson.D)) == 0 || documentKey.(bson.D)[0].Key != DefaultDocumentKeyColumnName {
+				// should never happen
+				return errors.New("invalid document key, expect _id")
+			}
+			id := documentKey.(bson.D)[0].Value
+			qValue, err := qValueStringFromKey(id)
+			if err != nil {
+				return fmt.Errorf("failed to convert _id to string: %w", err)
+			}
+			items.AddColumn(DefaultDocumentKeyColumnName, qValue)
+		} else {
+			// should never happen
+			return errors.New("documentKey field not found")
+		}
+
+		if fullDocument, found := changeDoc["fullDocument"]; found {
+			qValue, err := qValueJSONFromDocument(fullDocument.(bson.D))
+			if err != nil {
+				return fmt.Errorf("failed to convert fullDocument to JSON: %w", err)
+			}
+			items.AddColumn(DefaultFullDocumentColumnName, qValue)
+		} else {
+			if changeDoc["operationType"] == "delete" {
+				items.AddColumn(DefaultFullDocumentColumnName, types.QValueJSON{Val: "{}"})
+			} else {
+				// should never happen with fullDocument='UpdateLookup'
+				return errors.New("fullDocument field not found")
+			}
+		}
+
+		if operationType, ok := changeDoc["operationType"]; ok {
+			switch operationType {
+			case "insert":
+				if err := addRecord(ctx, &model.InsertRecord[model.RecordItems]{
+					BaseRecord:           model.BaseRecord{CommitTimeNano: clusterTimeNanos},
+					Items:                items,
+					SourceTableName:      sourceTableName,
+					DestinationTableName: destinationTableName,
+				}); err != nil {
+					return fmt.Errorf("failed to add insert record: %w", err)
+				}
+			case "update", "replace":
```

Comment:
looks like we're not going to really support update if I'm reading this right: update is for partial modifications, whereas replace is a complete update

Reply:
> update is for partial modifications, whereas replace is a complete update

correct. 
We are supporting both. Regardless whether the change event is update or replace, mongo will perform a fullDocument lookup after the change event. 

> is this something the user needs to configure?

User doesn't need to configure on their end, but we need to explicitly enable it in our code  [[ref](https://www.mongodb.com/docs/manual/changeStreams/#lookup-full-document-for-update-operations)]:
```
SetFullDocument(options.UpdateLookup)
```

The caveat is that unlike postgres/mysql, this result does not necessarily provide the document right after the update is applied, as there may be other writes that happened after the update but before the lookup.